src/proposals.rs

```diff
@@ -176,19 +206,65 @@ async fn poll_approved_proposals(
                 "proposal {} was None even though the query status code indicates it was found.",
                 proposal_id
             );
-            state.increment_proposal_id();
+
+            if pending_proposals {
```

Comment:
We're repeating this exact block in a lot of places, I wonder if we shouldn't just make a function out of it. You can make the pending proposal data part of the state too if that makes managing it easier, or just have it update mut vars in this function call.